### PR TITLE
Variables in metric instances

### PR DESCRIPTION
**Pull request by aaTman**
# EWB Pull Request

## Description

To help improve the API for EWB, the metrics now can have their own forecast and target variables declared in their instantiation.

Example: 

```python

...
# GHCN target
ghcn_target = inputs.GHCN(
    variables=["surface_air_temperature"],
)

# Define forecast (HRES)
hres_forecast = inputs.ZarrForecast(
    name="hres_forecast",
    source="gs://weatherbench2/datasets/hres/2016-2022-0012-1440x721.zarr",
    variables=["surface_air_temperature"],
    variable_mapping=inputs.HRES_metadata_variable_mapping,
)

metrics_list = [
    metrics.MaximumMAE(),
metrics.RMSE(forecast_variable='surface_eastward_wind',target_variable='surface_eastward_wind'),
    metrics.OnsetME(),
    metrics.DurationME(),
    metrics.MaxMinMAE(),
]
...
```

Where all of the metrics will evaluate the set of the variables in `ghcn_target` and `hres_forecast` (`surface_air_temperature`) except for `RMSE`, which will instead evaluate `surface_eastward_wind`.

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

New tests were set up to cover the core and edge cases of the variable permutations.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/extremeweatherbench/evaluate.py [hunk 1]
```diff
@@ -260,23 +260,39 @@ def compute_case_operator(
...
     logger.info(
         "Computing case operator for case %s...",
         case_operator.case_metadata.case_id_number,
     )
+
+    # Validate that all metrics are instantiated (not classes or callables)
+    for i, metric in enumerate(case_operator.metric_list):
+        if isinstance(metric, type):
+            metric = metric()
+            case_operator.metric_list[i] = metric
```
Comment (darothen) [inline src/extremeweatherbench/evaluate.py]: Not sure if this will be superseded by #228 but I had proposed the following:


```suggestion
            case_operator.metric_list[i] = metric()
```

File: src/extremeweatherbench/evaluate.py [hunk 2]
```diff
@@ -260,23 +260,39 @@ def compute_case_operator(
...
+            logger.warning(
+                "Metric %s instantiated with default parameters", metric.name
+            )
+        if not hasattr(metric, "compute_metric"):
+            raise TypeError(f"Metric must be a BaseMetric instance, got {type(metric)}")
+
```
Comment (darothen) [inline src/extremeweatherbench/evaluate.py]: Also potentially covered by #228...

```suggestion
        if not isinstance(metric, BaseMetric):
            raise TypeError(f"Metric must be a BaseMetric instance, got {type(metric)}")

```

File: src/extremeweatherbench/evaluate.py [hunk 3]
```diff
@@ -302,24 +318,36 @@ def compute_case_operator(
...
-                target_variable=variables[1],
-                metric=metric,
-                case_operator=case_operator,
-                **kwargs,
+
+    # Loop through metrics; if variables aren't defined for the metric, use variables
+    # declared in the InputBase objects
+    for metric in case_operator.metric_list:
+        # Determine which variable pairs to evaluate for this metric
+        if metric.forecast_variable is not None and metric.target_variable is not None:
```
Comment (darothen) [inline src/extremeweatherbench/evaluate.py]: nit: should add explicit grouping to each clause of the logical

```suggestion
        if (metric.forecast_variable is not None) and (metric.target_variable is not None):
```

**Review by darothen — APPROVED** [review]
